File: python_strategy/DojiCandleStrategy.py
from pandas import DataFrame

from entity import EventType, Candle
from strategy_v1.StopTargetEnum import StopTargetEnum
from strategy_v1.StrategyBase import StrategyBase


class DojiCandleStrategy(StrategyBase):
    def __init__(self, sl_target: StopTargetEnum):
        super().__init__(sl_target=sl_target)

    def analyze(self, df: DataFrame) -> tuple[EventType, str]:
        self.apply_indicator(df)

        candles  = Candle.get_candles(df, 3)
        cur_candle = candles[-1]
        prev_candle = candles[-2]
        pprev_candle = candles[-3]

        if prev_candle.is_doji(0.05):
            cur_row = df.iloc[-1]
            if cur_candle.is_bullish() and cur_row[self.atr_column] > (cur_candle.close - cur_candle.open) > cur_row[self.atr_column] * .5:
                return EventType.BUY, "Doji Candle"
            if cur_candle.is_bearish() and cur_row[self.atr_column] > (cur_candle.open - cur_candle.close) > cur_row[self.atr_column] * .5:
                return EventType.SELL, "Doji Candle"
        return EventType.NONE, ""

    def apply_indicator(self, candle_data: DataFrame):
        super().apply_indicator(candle_data)

File: src/strategy_v1/DojiCandleStrategy.js
import { StrategyBase } from './StrategyBase';

/**
 * Doji Candle Strategy
 * Identifies doji candles followed by strong directional moves
 */
export class DojiCandleStrategy extends StrategyBase {
    constructor(config = {}) {
        super(config);
        this.dojiThreshold = config.dojiThreshold || 0.05; // 5% body to range ratio
    }

    /**
     * Check if candle is a doji
     */
    isDoji(candle, threshold = 0.05) {
        const body = Math.abs(candle.close - candle.open);
        const range = candle.high - candle.low;
        
        if (range === 0) return false;
        
        const bodyRatio = body / range;
        return bodyRatio <= threshold;
    }

    /**
     * Check if candle is bullish
     */
    isBullish(candle) {
        return candle.close > candle.open;
    }

    /**
     * Check if candle is bearish
     */
    isBearish(candle) {
        return candle.close < candle.open;
    }

    /**
     * Analyze candle data
     */
    analyze(candleData) {
        if (candleData.length < 3 + this.atrPeriod) {
            return { signal: 'NONE', message: '', confidence: 0 };
        }

        const baseIndicators = this.applyIndicators(candleData);
        const atrIdx = baseIndicators.atr.length - 1;
        const currentATR = baseIndicators.atr[atrIdx];

        const currentCandle = candleData[candleData.length - 1];
        const prevCandle = candleData[candleData.length - 2];
        const pprevCandle = candleData[candleData.length - 3];

        // Check if previous candle is a doji
        if (this.isDoji(prevCandle, this.dojiThreshold)) {
            const currentBody = Math.abs(currentCandle.close - currentCandle.open);

            // Bullish signal: Doji followed by bullish candle
            if (this.isBullish(currentCandle)) {
                const bodyInRange = currentBody > currentATR * 0.5 && currentBody < currentATR;
                
                if (bodyInRange) {
                    return this.createSignal(
                        candleData,
                        'BUY',
                        'Doji Candle followed by bullish move',
                        0.7
                    );
                }
            }

            // Bearish signal: Doji followed by bearish candle
            if (this.isBearish(currentCandle)) {
                const bodyInRange = currentBody > currentATR * 0.5 && currentBody < currentATR;
                
                if (bodyInRange) {
                    return this.createSignal(
                        candleData,
                        'SELL',
                        'Doji Candle followed by bearish move',
                        0.7
                    );
                }
            }
        }

        return { signal: 'NONE', message: '', confidence: 0 };
    }
}
